File: solutions/python/triangle/1/triangle.py
def is_triangle(sides):
    """Check if three sides can form a valid triangle."""
    a, b, c = sides
    # All sides must be positive
    if a <= 0 or b <= 0 or c <= 0:
        return False
    # Triangle inequality theorem
    return (a + b >= c) and (b + c >= a) and (a + c >= b)


def equilateral(sides):
    """Check if triangle is equilateral (all sides equal)."""
    if not is_triangle(sides):
        return False
    a, b, c = sides
    return a == b == c


def isosceles(sides):
    """Check if triangle is isosceles (at least two sides equal)."""
    if not is_triangle(sides):
        return False
    a, b, c = sides
    return a == b or b == c or a == c


def scalene(sides):
    """Check if triangle is scalene (all sides different)."""
    if not is_triangle(sides):
        return False
    a, b, c = sides
    return a != b and b != c and a != c

File: solutions/cpp/triangle/1/triangle.cpp
#include "triangle.h"

namespace triangle {
    flavor kind(double a, double b, double c) {
        // All sides must be positive
        if (a <= 0 || b <= 0 || c <= 0) {
            throw std::domain_error("Sides must be positive");
        }
        // Triangle inequality: sum of any two sides must be >= the third
        if (a + b < c || a + c < b || b + c < a) {
            throw std::domain_error("Invalid triangle: violates triangle inequality");
        }
        // Determine type
        if (a == b && b == c) {
            return flavor::equilateral;
        }
        if (a == b || a == c || b == c) {
            return flavor::isosceles;
        }
        return flavor::scalene;
    }
}